server/src/ingestion.rs

```diff
@@ -0,0 +1,135 @@
+            client_id,
+            topics,
+            qos,
+        }
+    }
+
+    /// Create a broker client, subscribe to the topics, and start ingesting messages.
+    pub fn start(self, mut storage_engine: StorageEngine) {
+        println!("Creating MQTT broker client.");
+        let mut client = self.create_client();
+
+        if let Err(err) = block_on(async {
+            let mut stream = self.subscribe_to_broker(&mut client).await;
+
+            println!("Waiting for messages...");
+            Self::ingest_messages(&mut stream, &mut client, &mut storage_engine).await;
+
+            Ok::<(), mqtt::Error>(())
+        }) {
+            eprintln!("{}", err);
+        }
+    }
+
+    /// Create a broker client with the ingestors broker URI and client ID.
+    fn create_client(&self) -> AsyncClient {
+        let create_options = mqtt::CreateOptionsBuilder::new()
+            .server_uri(self.broker)
+            .client_id(self.client_id)
+            .finalize();
+
+        mqtt::AsyncClient::new(create_options).unwrap_or_else(|e| {
+            eprintln!("Error creating the client: {:?}", e);
+            process::exit(1);
+        })
+    }
+
+    /// Make the connection to the broker and subscribe to the specified topics.
+    async fn subscribe_to_broker(
+        &self,
+        client: &mut AsyncClient,
+    ) -> AsyncReceiver<Option<Message>> {
+        // Get message stream before connecting since messages can arrive as soon as the connection is made.
+        let mut stream = client.get_stream(25);
+
+        // Define last will and testament message to notify other clients about disconnect.
+        let lwt = mqtt::Message::new("mdb_lwt", "ModelarDB lost connection", mqtt::QOS_1);
+
+        let connect_options = mqtt::ConnectOptionsBuilder::new()
+            .keep_alive_interval(Duration::from_secs(30))
+            .mqtt_version(mqtt::MQTT_VERSION_3_1_1)
+            .clean_session(false)
+            .will_message(lwt)
+            .finalize();
+
+        println!("Connecting to MQTT broker.");
+        client.connect(connect_options).await;
+
+        println!("Subscribing to topics: {:?}", self.topics);
+        client.subscribe_many(self.topics, self.qos).await;
+
+        stream
+    }
+
+    /// Ingest the published messages in a loop until connection is lost.
+    async fn ingest_messages(
+        stream: &mut AsyncReceiver<Option<Message>>,
+        client: &mut AsyncClient,
+        storage_engine: &mut StorageEngine,
+    ) {
+        // While the message stream resolves to the next item in the stream, ingest the messages.
+        while let Some(msg_opt) = stream.next().await {
+            if let Some(msg) = msg_opt {
+                storage_engine.insert_message(msg);
+            } else {
+                // A "None" means we were disconnected. Try to reconnect.
+                println!("Lost connection. Attempting reconnect.");
+                while let Err(err) = client.reconnect().await {
```

Comment:
It is unclear how this loop ever terminates if `client` is never able to reconnect.

Reply:
Removed the reconnect feature. As mentioned earlier, it might no be relevant to spend time on more "advanced" functionality such as this since this component will be replaced in the near future

Reply:
I agree.